Generic/CyclomaticComplexity: improve code coverage

rodrigoprimo opened the pull request:
# Description

This PR improves code coverage for the `Generic.Metrics.CyclomaticComplexity` sniff.

I noticed three things while working on this sniff that we might want to address in the future:

- The sniff does not check the cyclomatic complexity of anonymous functions, and I believe it should. That being said, I'm not too familiar with the concept of cyclomatic complexity, and I could be missing something.
- The sniff error message could be improved if it included the function name.
- There are some divergences between how PHPCS and PHPMD/PHP Depend count cyclomatic complexity. For example, PHPCS increases the cyclomatic complexity by one whenever it finds a `default` statement, while PHPMD/PHP Depend doesn't (https://pdepend.org/documentation/software-metrics/cyclomatic-complexity.html).

Just looking for some initial feedback on those three items to decide if I should address them or not on future PRs. I am happy to move them to an issue as well if you prefer. I'm inclined to think that the first two items are worthwhile and the third is not (assuming this is not a very popular sniff and that there is no easy-to-find clear definition of how to count cyclomatic complexity in PHP).

## Related issues/external references

Part of https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/146

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

<!--
============================================================================================
Please make sure your pull request passes all continuous integration checks!

PRs which are failing their CI checks will likely be ignored by the maintainers.

Small PRs using atomic, descriptive commits are hugely appreciated as it will make
reviewing your changes easier for the maintainers.
============================================================================================
-->

jrfnl commented:
> Just looking for some initial feedback on those three items to decide if I should address them or not on future PRs. I am happy to move them to an issue as well if you prefer.

Please move these questions/points to individual issues. They each have their own decision point and those decision points are unrelated to the decision point of this PR.

jrfnl reviewed (changes requested):
@rodrigoprimo Thanks for this PR and for continuing with the work to increase code coverage.

IMO this one needs a little more work.

Aside from the comments left in-line, I'd like to ask you to have a look at the list of tokens the sniff is looking for:

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/da0f04b5432a3adec175f9352a059058a1557533/src/Standards/Generic/Sniffs/Metrics/CyclomaticComplexitySniff.php#L73-L87


I can see at least four tokens in that list which are not used in the tests. I believe tests should be added/adjusted to ensure those tokens are included in the tests.


Aside from that, on top of the questions you already asked (and which need to be moved to issues), I also saw another issue in this sniff for which I've opened issue #726.

rodrigoprimo commented:
> Please move these questions/points to individual issues. They each have their own decision point and those decision points are unrelated to the decision point of this PR.

I just created the three issues:

https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/742
https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/743
https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/744

rodrigoprimo commented on src/Standards/Generic/Sniffs/Metrics/CyclomaticComplexitySniff.php:
```diff
@@ -60,7 +60,7 @@ public function process(File $phpcsFile, $stackPtr)
 
-        // Ignore abstract methods.
+        // Ignore abstract and interface methods. Bail early when live coding.
         if (isset($tokens[$stackPtr]['scope_opener']) === false) {
```
My understanding is that it only makes sense to run this sniff when the function has a scope opener and a scope closer so that it knows when to start counting cyclomatic complexity and when it should stop.

That being said, as far as I can test, `scope_opener` and `scope_closer` are only set when both are present in the code. So it seems to me that it is fine to check just for `scope_opener` as it is done in this sniff.

Do you agree? I have a vague memory of us discussing something similar in the past, but I don't recall the details, and I could not find it.

I checked and saw an example of a sniff that use `scope_closer`, but only check if `scope_opener` is set and an example of a sniff that check for both:

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/9bf4e15b9ddfc16d162670702f54dc09fb0345a1/src/Standards/Generic/Sniffs/PHP/LowerCaseTypeSniff.php#L94-L98

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/84acf4e56f110db8e75cb9a575c5727df637643c/src/Standards/Generic/Sniffs/Metrics/NestingLevelSniff.php#L60-L66

rodrigoprimo commented on src/Standards/Generic/Tests/Metrics/CyclomaticComplexityUnitTest.2.inc:
```diff
@@ -0,0 +1,7 @@
+// This should be the only test in this file.
+// Testing that the sniff is *not* triggered.
+
+function sniffShouldBailMissingScopeOpener()
```
I think so to ensure that the sniff is not triggered in this case. Thanks for mentioning it. I just added it.

rodrigoprimo commented:
Thanks for your review, @jrfnl. I just replied to your inline comments.

> Aside from the comments left in-line, I'd like to ask you to have a look at the list of tokens the sniff is looking for:
> I can see at least four tokens in that list which are not used in the tests. I believe tests should be added/adjusted to ensure those tokens are included in the tests.

Good point. I updated the existing tests to include at least one occurrence of each of the tokens that the sniff looks for to count cyclomatic complexity. I added the following tokens: T_ELSEIF, T_CATCH, T_FOREACH and T_FOR.

jrfnl reviewed (approved):
Thanks again for this @rodrigoprimo !

I'll rebase the PR and squash most of the commits. Will merge once the build has passed again.
